012_ext-xmlrpc-bad-callback-crash.patch
PHP_5_2
http://cvs.php.net/viewvc.cgi/php-src/ext/xmlrpc/xmlrpc-epi-php.c?r1=1.39.2.5.2.13&r2=1.39.2.5.2.14&diff_format=u
http://cvs.php.net/viewvc.cgi/php-src/ext/xmlrpc/tests/bug47818.phpt?view=markup&revision=1.1
http://bugs.php.net/bug.php?id=47818
Fixed bug #47818 (Segfault due to bound callback param)

diff -r a5eccb8408e8 -r f582e5464d52 ext/xmlrpc/tests/bug47818.phpt
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/ext/xmlrpc/tests/bug47818.phpt	Fri Apr 10 11:42:59 2009 +0200
@@ -0,0 +1,39 @@
+--TEST--
+Bug #47818 (Segfault due to bound callback param)
+--FILE--
+<?php
+
+class MyXmlRpc {
+    private $s;
+    private $method;
+
+    function impl($method_name, $params, $user_data){
+        $this->method = $method_name;
+        print "Inside impl(): {$this->method}\n";
+        return array_sum($params);
+    }
+
+    function __construct() {
+        $this->s = xmlrpc_server_create();
+        xmlrpc_server_register_method($this->s, 'add', array($this, 'impl'));
+    }
+
+    function call($req) {
+        return xmlrpc_server_call_method($this->s, $req, null);
+    }
+
+    function getMethod() {return $this->method;}
+
+}
+
+$x = new MyXmlRpc;
+$resp = $x->call(xmlrpc_encode_request('add', array(1, 2, 3)));
+
+$method = $x->getMethod();
+
+print "Global scope: $method\n";
+
+?>
+--EXPECTF--
+Inside impl(): add
+Global scope: add
diff -r a5eccb8408e8 -r f582e5464d52 ext/xmlrpc/xmlrpc-epi-php.c
--- a/ext/xmlrpc/xmlrpc-epi-php.c	Fri Apr 10 11:41:21 2009 +0200
+++ b/ext/xmlrpc/xmlrpc-epi-php.c	Fri Apr 10 11:42:59 2009 +0200
@@ -1113,8 +1113,8 @@
 			}
 
 			/* cleanup after ourselves.  what a sty! */
-			zval_dtor(data.xmlrpc_method);
-			FREE_ZVAL(data.xmlrpc_method);
+			zval_ptr_dtor(&data.xmlrpc_method);
+
 			zval_dtor(data.return_data);
 			FREE_ZVAL(data.return_data);
 
